### ISSUE TITLE
JSON-RPC issues on 'next' branch

### ISSUE THREAD

**Issue by craigraw** [romanz/electrs#310]
A few JSON-RPC issues I've encountered working on the `next` branch:

1. Strings are not accepted as IDs, although they are valid per the JSON-RPC spec. It's not a significant problem to change to integers, but it might affect some clients.
2. The second `verbose` parameter is not accepted on a `blockchain.transaction.get` command. This command is important because in the general case (i.e. without a script hash) you cannot retrieve the block height for a txid to know whether it has been confirmed or not. Happy to hear if there is a substitute way of doing this that is more efficient for the server.
3. The error `failed: no subscription for scripthash` is received on a `blockchain.scripthash.get_history` call for a script hash that has not been subscribed. The use case for this situation is trying to get the block height for a txid when the `verbose` parameter is not accepted but the script hash is known, so this may be avoided by solving 2. above. It is a little unexpected though.

**Comment by romanz** [comment]
Thanks for reporting these issues :)

**Comment by romanz** [comment]
(1) should be resolved by 96e3e6222e95c7e35e639ca87cfc8c83c52c3a7c.

**Comment by romanz** [comment]
(2) is a bit tricky to implement for "non-subscribed" transactions, since the new code doesn't build a txindex.
The idea is that users usually query the txids they got via `blockchain.scripthash.subscribe` notifications.
Those transactions (and their block hashes) are cached in-memory - so it is possible to call bitcoind `getrawtransaction` RPC with its block hash.

**Comment by craigraw** [comment]
This is understandable after inspecting the bitcoind `getrawtransaction` RPC more carefully. It's possible to subscribe the associated script hash first in some situations, so I will try that. One thing I should have noted though is that specifying the `verbose` parameter causes the server to crash with `Error: invalid length 2, expected fewer elements in array` - this means it's difficult to test whether verbose works before reverting to subscribing the script hash and querying the txid.

Is there a way that Electrs could support the general use case when the bitcoind node has `txindex=1`? While it's not the bitcoind default, it is increasingly the default for packaged nodes and install guides given the wide need for it across other applications. The new `getindexinfo` RPC can return the status of txindex: https://github.com/bitcoin/bitcoin/pull/19550

**Comment by romanz** [comment]
> (2) is a bit tricky to implement for "non-subscribed" transactions, without building a txindex.
> The idea is that users usually query the txids they got via `blockchain.scripthash.subscribe` notifications.
> Those transactions (and their block hashes) are cached in-memory - so it is possible to call bitcoind `getrawtransaction` RPC with its block hash.

Should be resolved by https://github.com/romanz/electrs/commit/cb65920224220eeeee34895615519506f045ca01.
Please make sure to reindex the blockchain :)

**Comment by romanz** [comment]
>     2\. The second `verbose` parameter is not accepted on a `blockchain.transaction.get` command.

It should now work with fd14a83 :)

**Comment by craigraw** [comment]
Confirmed both the verbose and non-verbose `blockchain.transaction.get` calls are working without subscriptions. Thanks!

Some notes on timing: The testnet re-index too about 3 hours, which seems very reasonable. After subscribing about 400 script hashes, a new block indexing takes about 25 seconds on my RPi4. It appears no RPC calls are answered while this is taking place, but (importantly) Electrs does respond to them after the indexing is completed. I've logged some times to the verbose `blockchain.transaction.get` call in milliseconds - the longer one was initiated during an index:
```
2020-11-02 10:10:35,862 INFO Time: 173
2020-11-02 10:10:35,913 INFO Time: 51
2020-11-02 10:13:41,206 INFO Time: 13408
2020-11-02 10:13:41,265 INFO Time: 59
```

CPU usage is very low except when indexing a new block, even with this many script hashes subscribed. This is a dramatic and very welcome change from the 0.8.6 version. 

The only remaining JSON-RPC issue I had was in issuing a `blockchain.scripthash.get_history` call to a script hash that was not subscribed. This is useful in a blockchain explorer use case to traverse 'forward' in the transaction graph, finding out whether a particular output has been spent and by what transaction. I understand why this occurs, and it's not difficult to issue a `blockchain.scripthash.subscribe` call before these queries, but I thought I would mention it as it does add some additional load to the server by subscribing script hashes that do not need to be notified.

Thanks again!
